### ISSUE TITLE
Exception in thread "main" java.lang.UnsatisfiedLinkError: no jniUVCController in java.library.path

### ISSUE THREAD

**Issue by nospam2k** [CyberCastle/jLibuvc#9]
Not sure what to do about this.

**Comment by CyberCastle** [comment]
Hi

Can you share your code to replicate the issue, please?

Greetings

**Comment by nospam2k** [comment]
All I did was download the zip and open it in intellij and create a configuration adding only the main class from run and no other changes. The error is coming from:

```
   static {
        Loader.load();
    }
```

Here is the full console:

```
/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/bin/java -javaagent:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=61012:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/tools.jar:/Users/david/Downloads/jLibuvc-master/target/classes:/Users/david/.m2/repository/org/bytedeco/javacpp/1.0/javacpp-1.0.jar cl.cc.example.Run
Exception in thread "main" java.lang.UnsatisfiedLinkError: no jniUVCController in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:550)
	at org.bytedeco.javacpp.Loader.load(Loader.java:415)
	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
	at cl.cc.jlibuvc.UVCController.<clinit>(UVCController.java:122)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:390)
	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
	at cl.cc.jlibuvc.UVCController$UVCStreamCtrl.<clinit>(UVCController.java:276)
	at cl.cc.example.Run.main(Run.java:27)

Process finished with exit code 1
```

**Comment by CyberCastle** [comment]
Questions:

What is your OS?
Did you install libusb?

This software uses Javacpp to generate a JNI pointed to libusb, therefore this library must be installed on the system.
The error is because the mentioned JNI class is not found.

**Comment by nospam2k** [comment]
MacOS Catalina. I installed libusb with brew.

**Comment by nospam2k** [comment]
Ok, so on a different machine. I installed using brew libusb and libuvc. I'm using IntelliJ Idea IDE. I still get the same error. In the UVCController:

```
@Properties({
    @Platform(include = {"libuvc/libuvc.h"})})
public class UVCController implements InfoMapper {
```

Do I need to add anything to the Configuration to define where libuvc.h is? I tried just changing to the direct path (/usr/local/Cellar/libuvc/0.0.6/include/libuvc/libuvc.h) but still get the same error. Any help you can give me on this would be greatly appreciated.

NOTE: I'm just running this in the IDE. I haven't created a jar.

**Comment by CyberCastle** [comment]
I'm updating this project, for reply this issue and find a fix.

**Comment by nospam2k** [comment]
Sounds good!

**Comment by nospam2k** [comment]
Just an FYI. I was able to compile using maven once I changed the includePath and the linkPath (I don't normally use maven from the command line as I usually compile in the IDE). I'm not exactly sure how to translate this into the IDE. I'm assuming it will be in the VM options for the configuration. I also tried putting the javacpp-1.0.jar file as a library but still having the same error. Let me know if you have any thoughts. I will let you know if I find anything more.

**Comment by CyberCastle** [comment]
Please:

Update your local repository and execute:

`mvn clean package`
`java -Dsun.java2d.opengl=True -jar target/dist/jLibUVC.jar`

Let me know if the demo run successfully.

**Comment by nospam2k** [comment]
/Users/david/Src/jLibuvc/target/classes/cl/cc/jlibuvc/jniUVCController.cpp:5294:20: error: use of undeclared identifier 'uvc_mjpeg2rgb'
        int rval = uvc_mjpeg2rgb(ptr0, ptr1);
                   ^
1 error generated.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.791 s
[INFO] Finished at: 2020-07-07T21:56:50-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.bytedeco:javacpp:1.5.3:build (process-classes) on project jLibUVC: Execution process-classes of goal org.bytedeco:javacpp:1.5.3:build failed: Process exited with an error: 1 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

**Comment by CyberCastle** [comment]
Please install libjpeg, since this library is an optional libuvc dependency:

`brew install libjpeg`

**Comment by nospam2k** [comment]
Libjpeg was installed. I reinstalled, brew doctor, brew upgrade but still same error.

**Comment by CyberCastle** [comment]
I don't understand why you can't compile the code. Please try commenting this line

<img width="1155" alt="image" src="https://user-images.githubusercontent.com/905565/86950820-aafb5880-c11e-11ea-8959-05aa15ecbb9d.png">

In file UVCController.java

**Comment by nospam2k** [comment]
Still failed (see below) couldn't find symbol. I did a grep -r uvc_mjpeg2rgb *:

```
0.0.6/include/libuvc/libuvc.h:uvc_error_t uvc_mjpeg2rgb(uvc_frame_t *in, uvc_frame_t *out);
```

library is in /usr/local/Cellar/libuvc (alias /usr/local/opt)

versions:
libuvc 0.0.6
libusb 1.0.23
jpeg 9d


(FYI: I'm free all day to test this and can test immediately, if that helps you)

```
$ mvn clean package
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------------< cl.cc:jLibUVC >----------------------------
[INFO] Building jLibUVC 1.2
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ jLibUVC ---
[INFO] Deleting /Users/david/Src/jLibuvc/target
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-versions) @ jLibUVC ---
[INFO] 
[INFO] --- maven-dependency-plugin:3.1.2:properties (default) @ jLibUVC ---
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ jLibUVC ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/david/Src/jLibuvc/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ jLibUVC ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 4 source files to /Users/david/Src/jLibuvc/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/david/Src/jLibuvc/src/main/java/cl/cc/Run.java:[85,48] cannot find symbol
  symbol:   method uvc_mjpeg2rgb(cl.cc.jlibuvc.UVCController.UVCFrame,cl.cc.jlibuvc.UVCController.UVCFrame)
  location: class cl.cc.jlibuvc.UVCController
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.580 s
[INFO] Finished at: 2020-07-08T10:33:41-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jLibUVC: Compilation failure
[ERROR] /Users/david/Src/jLibuvc/src/main/java/cl/cc/Run.java:[85,48] cannot find symbol
[ERROR]   symbol:   method uvc_mjpeg2rgb(cl.cc.jlibuvc.UVCController.UVCFrame,cl.cc.jlibuvc.UVCController.UVCFrame)
[ERROR]   location: class cl.cc.jlibuvc.UVCController
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

**Comment by nospam2k** [comment]
I think this is where the compile problem lies after remming out. In Run.java

I can compile if I comment out the following:

```
        final DemoFrame windows = new DemoFrame("Sonria!!!!!! :-)");
        windows.setSize(1280, 720);
        UVCFrameCallback callback = new UVCFrameCallback() {

            @Override
            public void call(UVCFrame frame) {

                /* Commented out
                try {
                    UVCFrame cframe = UVCController.uvc_allocate_frame(frame.width() * frame.height());
                    int convres = UVCController.uvc_mjpeg2rgb(frame, cframe);
                    if (convres != 0) {
                        return;
                    }

                    windows.showImage(Utils.getImage(cframe));

                    UVCController.uvc_free_frame(cframe);
                } catch (Exception e) {
                    e.printStackTrace(System.out);
                }
                */
            }

        };
```

**Comment by nospam2k** [comment]
Running `java -Djava.library.path=./lib -jar jLibUVC.jar` from the dist dir gives the error: 

```
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Init Result: 0
Find Result: -4
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010eb44cd8, pid=40375, tid=0x0000000000001003
#
# JRE version: Java(TM) SE Runtime Environment (8.0_231-b11) (build 1.8.0_231-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.231-b11 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libuvc.dylib+0x4cd8]  uvc_open+0x17
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/david/Src/jLibuvc/target/dist/hs_err_pid40375.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by CyberCastle** [comment]
Sorry, my fault. My camera is a Logitech 920, so it needs to use libjpeg to get the image. I dont'n idea how test the code without that library.

The code that you have commented, allows to visualize the frames coming from the camera. 

The "Utils.getImage" method allows you to convert an RGB frame to a BufferedImage object so that it can be viewed.

You should investigate in what format the camera frames come and convert it to a BufferedImage object.

**Comment by nospam2k** [comment]
Ok, so for now, can we just work with using HD Camera so we can compare problems? Once we can get past working with that, I can troubleshoot specifics.

My ultimate goal is to use this on Windows 10 and had problems compiling, but I didn't want to confuse things.

**Comment by CyberCastle** [comment]
HD Camera isn't a UVC device, you need a UVC camera to be able to use this code.

**Comment by nospam2k** [comment]
Ah, ok, I do have a UVC camera, I didn't know that the built in was not. Ok, I'll report back.

**Comment by nospam2k** [comment]
> The "Utils.getImage" method allows you to convert an RGB frame to a BufferedImage object so that it can be viewed.
> 
> You should investigate in what format the camera frames come and convert it to a BufferedImage object.

Question: I'm a bit out of my league on this so I need your help as to where I'm making this change. My camera supports h.264 and mjpeg (Aver CAM520).

I'm on a different mac with just the current git clone of the repo, libjpeg, libusb, libuvc installed, the CAM520 is connected. I'm getting the same undeclared identifier error as at first. Are you sure this is camera related? It seems to be complaining about a declaration.

**Comment by nospam2k** [comment]
I changed the call I remmed out earlier to:

```
 try {
                    UVCFrame cframe = UVCController.uvc_allocate_frame(frame.width() * frame.height());
                    //int convres = UVCController.uvc_mjpeg2rgb(frame, cframe); <<< commented out for testing
                    int convres = 1; // <<< added for testing
                    if (convres != 0) {
                        return;
                    }

                    windows.showImage(Utils.getImage(cframe));

                    UVCController.uvc_free_frame(cframe);
                } catch (Exception e) {
                    e.printStackTrace(System.out);
                }
```

and I still get the same error so it seems that maybe something related to:

```
public static native int uvc_mjpeg2rgb(UVCFrame fin, UVCFrame fout);

```

UPDATE: As a test, I have commented out the line from UVCController.java and changed uvc_mjpeg2rgb to uvc_any2rgb and that compiled ok. (I'm really just guessing, I'm a hacker, not a programmer ;))

**Comment by nospam2k** [comment]
Progress! The Sonria window came up, but blank and this error.

```
java -jar jLibUVC.jar 
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Init Result: 0
Find Result: 0
unsupported descriptor subtype VS_COLORFORMAT
unsupported descriptor subtype VS_COLORFORMAT
unsupported descriptor subtype VS_COLORFORMAT
Open Result: 0
Warning: Could not load PointerPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
0
Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Serial Number: 5309643100040
Manufacturer: org.bytedeco.javacpp.BytePointer[address=0x7fe673e2e7c0,position=0,limit=0,capacity=0,deallocator=null]
Product Description: CAM520
Descriptor Subtype: 4
Bits Per Pixel: 16
Aspect Ratio X: 0
Aspect Ratio Y: 0
Default Frame Index: 6
Frame Descriptors: 0
Interlace Flags: 0
Get Stream Control Result: 0
Frame Index: 12
Max Video Frame Buffer Size: 1843200
Frame Interval: 333333
Interface Number: 1
attempt to claim already-claimed interface 1
Error creating global reference of cl.cc.jlibuvc.UVCController.UVCFrameCallback instance for callback.
Function pointer object is NULL in callback for cl.cc.jlibuvc.UVCController.UVCFrameCallback.
Exception in thread "JavaCPP Thread ID 67083" java.lang.InstantiationException: cl.cc.jlibuvc.UVCController$UVCFrameCallback
...
(repeats Above error)
```

I imagine I'm missing something on the command line(?)

**Comment by nospam2k** [comment]
I see this in the libuvc.h

```
#ifdef LIBUVC_HAS_JPEG
uvc_error_t uvc_mjpeg2rgb(uvc_frame_t *in, uvc_frame_t *out);
#endif
```

Is it possible that brew isn't compiling that?

**Comment by nospam2k** [comment]
Ok, using compiled libuvc allows for compile without any need for mods, but crashes with this:
```
java -jar target/dist/jLibUVC.jar 
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Init Result: 0
Find Result: 0
Open Result: 0
Warning: Could not load PointerPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
0
Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Serial Number: 5309643100040
Manufacturer: org.bytedeco.javacpp.BytePointer[address=0x7faa24c2d880,position=0,limit=0,capacity=0,deallocator=null]
Product Description: CAM520
Descriptor Subtype: 4
Bits Per Pixel: 16
Aspect Ratio X: 0
Aspect Ratio Y: 0
Default Frame Index: 6
Frame Descriptors: 0
Interlace Flags: 0
Get Stream Control Result: 0
Frame Index: 12
Max Video Frame Buffer Size: 1843200
Frame Interval: 333333
Interface Number: 1
```

Pertinent (I think) part from crash log that pops up.

```
Thread 40 Crashed:
0   libsystem_kernel.dylib        	0x00007fff6869433a __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff68750e60 pthread_kill + 430
2   libsystem_c.dylib             	0x00007fff6861b808 abort + 120
3   libc++abi.dylib               	0x00007fff6587b458 abort_message + 231
4   libc++abi.dylib               	0x00007fff6586c8a7 demangling_terminate_handler() + 238
5   libobjc.A.dylib               	0x00007fff673a75b1 _objc_terminate() + 104
6   libc++abi.dylib               	0x00007fff6587a887 std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff6587d1a2 __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 27
8   libc++abi.dylib               	0x00007fff6587d169 __cxa_throw + 113
here >>> 9   libjniUVCController.dylib     	0x00000001153ed124 JavaCPP_cl_cc_jlibuvc_UVCController_00024UVCFrameCallback::operator()(uvc_frame*, void*) + 1076
10  libuvc.0.0.6.dylib            	0x000000010f5eb61f _uvc_user_caller + 127
11  libsystem_pthread.dylib       	0x00007fff68751109 _pthread_start + 148
12  libsystem_pthread.dylib       	0x00007fff6874cb8b thread_start + 15
```

**Comment by CyberCastle** [comment]
> > The "Utils.getImage" method allows you to convert an RGB frame to a BufferedImage object so that it can be viewed.
> > You should investigate in what format the camera frames come and convert it to a BufferedImage object.
> 
> Question: I'm a bit out of my league on this so I need your help as to where I'm making this change. My camera supports h.264 and mjpeg (Aver CAM520).
> 
> I'm on a different mac with just the current git clone of the repo, libjpeg, libusb, libuvc installed, the CAM520 is connected. I'm getting the same undeclared identifier error as at first. Are you sure this is camera related? It seems to be complaining about a declaration.

No, this issue leading to this thread is because for some strange reason JavaCPP cannot find libjpeg library when generating jni dependency.

> I see this in the libuvc.h
> 
> ```
> #ifdef LIBUVC_HAS_JPEG
> uvc_error_t uvc_mjpeg2rgb(uvc_frame_t *in, uvc_frame_t *out);
> #endif
> ```
> 
> Is it possible that brew isn't compiling that?

Right, that's the problem.

Please, update your local repository (for clean some warnings) and try installing libuvc from source:

`brew install libuvc --HEAD`

I'm running this project on macOS Mojave, with Java 14.0.

**Comment by nospam2k** [comment]
Ahhhhh.... OK, I have been avoiding upgrading java past 8. I just installed 14 and libuvc --HEAD.

It's working but crashed when I click away from the window.

I still get some Warnings:

```
java -jar target/dist/jLibUVC.jar 
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
Init Result: 0
Find Result: 0
Open Result: 0
Warning: Could not load PointerPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
0
Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
Serial Number: 5309643100040
Manufacturer: org.bytedeco.javacpp.BytePointer[address=0x7fa9647360f0,position=0,limit=0,capacity=0,deallocator=null]
Product Description: CAM520
Descriptor Subtype: 4
Bits Per Pixel: 16
Aspect Ratio X: 0
Aspect Ratio Y: 0
Default Frame Index: 6
Frame Descriptors: 0
Interlace Flags: 0
Get Stream Control Result: 0
Frame Index: 12
Max Video Frame Buffer Size: 1843200
Frame Interval: 333333
Interface Number: 1
```

**Comment by nospam2k** [comment]
Also,
  Trying to move over to using Intellij Idea to debug I get:

```
/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/bin/java -javaagent:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=50981:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/tools.jar:/Users/david/Downloads/jLibuvc-master/target/classes:/Users/david/Downloads/jLibuvc-master/lib/javacpp-1.0.jar:/Users/david/.m2/repository/org/bytedeco/javacpp/1.0/javacpp-1.0.jar cl.cc.example.Run
Exception in thread "main" java.lang.UnsatisfiedLinkError: no jniUVCController in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:550)
	at org.bytedeco.javacpp.Loader.load(Loader.java:415)
	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
	at cl.cc.jlibuvc.UVCController.<clinit>(UVCController.java:122)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:390)
	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
	at cl.cc.jlibuvc.UVCController$UVCStreamCtrl.<clinit>(UVCController.java:276)
	at cl.cc.example.Run.main(Run.java:27)

Process finished with exit code 1
```

**Comment by CyberCastle** [comment]
> Ahhhhh.... OK, I have been avoiding upgrading java past 8. I just installed 14 and libuvc --HEAD.
> 
> It's working but crashed when I click away from the window.
> 
> I still get some Warnings:
> 
> ```
> java -jar target/dist/jLibUVC.jar 
> Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
> Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
> Init Result: 0
> Find Result: 0
> Open Result: 0
> Warning: Could not load PointerPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
> 0
> Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/Users/david/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
> Serial Number: 5309643100040
> Manufacturer: org.bytedeco.javacpp.BytePointer[address=0x7fa9647360f0,position=0,limit=0,capacity=0,deallocator=null]
> Product Description: CAM520
> Descriptor Subtype: 4
> Bits Per Pixel: 16
> Aspect Ratio X: 0
> Aspect Ratio Y: 0
> Default Frame Index: 6
> Frame Descriptors: 0
> Interlace Flags: 0
> Get Stream Control Result: 0
> Frame Index: 12
> Max Video Frame Buffer Size: 1843200
> Frame Interval: 333333
> Interface Number: 1
> ```
Warnings disappear if you have updated your local repository.








> Also,
> Trying to move over to using Intellij Idea to debug I get:
> 
> ```
> /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/bin/java -javaagent:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/lib/idea_rt.jar=50981:/Users/david/Applications/IntelliJ IDEA CE.app/Contents/bin -Dfile.encoding=UTF-8 -classpath /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/lib/tools.jar:/Users/david/Downloads/jLibuvc-master/target/classes:/Users/david/Downloads/jLibuvc-master/lib/javacpp-1.0.jar:/Users/david/.m2/repository/org/bytedeco/javacpp/1.0/javacpp-1.0.jar cl.cc.example.Run
> Exception in thread "main" java.lang.UnsatisfiedLinkError: no jniUVCController in java.library.path
> 	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
> 	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
> 	at java.lang.System.loadLibrary(System.java:1122)
> 	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:550)
> 	at org.bytedeco.javacpp.Loader.load(Loader.java:415)
> 	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
> 	at cl.cc.jlibuvc.UVCController.<clinit>(UVCController.java:122)
> 	at java.lang.Class.forName0(Native Method)
> 	at java.lang.Class.forName(Class.java:348)
> 	at org.bytedeco.javacpp.Loader.load(Loader.java:390)
> 	at org.bytedeco.javacpp.Loader.load(Loader.java:358)
> 	at cl.cc.jlibuvc.UVCController$UVCStreamCtrl.<clinit>(UVCController.java:276)
> 	at cl.cc.example.Run.main(Run.java:27)
> 
> Process finished with exit code 1
> ```
Sorry i don't use that ide so i can't help you.

**Comment by nospam2k** [comment]
Ok, so I got the IDE worked out. I just needed to at the jLibUVC.jar to the Libraries in the project structure. I'm off to the races! Thanks for the help. The window switch crash is still happening but I'm not going to be using the Run.java anyway and I'm only making a PTZ controller so I won't need video.

I will be trying to port this over to Windows once I get it working on Mac so I be back to either report good or look for trouble shooting but, for me, this is closable

Thanks for the help and the great library!

**Comment by nospam2k** [comment]
Little problem (didn't want to open another issue)

```
 System.out.println("Init Result: " + UVCController.uvc_init(ctx, null));
 PointerPointer<UVCController.UVCDevice> camList = new PointerPointer<>(new UVCController.UVCDevice());
 System.out.println("Get cameras:" + UVCController.uvc_get_device_list(ctx, camList));
 ```

Just trying to get a camera list. This crashes. I'm not completely clear on all the structures and how they work. Appreciate the help!

**Comment by nospam2k** [comment]
nm. Got it.

**Comment by CyberCastle** [comment]
Closing this thread
